app/models/spree/payment_method_decorator.rb

```diff
@@ -23,6 +23,11 @@
     end
   }
 
+  scope :for_distributors, ->(distributors) {
+    non_unique_matches = unscoped.joins(:distributors).where(enterprises: { id: distributors })
+    where(id: non_unique_matches.map(&:id))
```

Comment:
I don't understand this. In line 28 we execute the query built in line 27, pick up the ids from its result and then querying them again and so getting the same results of line 27 again.

Reply:
The `joins` here means that you could end up with duplicates. So, this is a workaround to avoid having to use `DISTINCT` or `group` for each query and just have a simple and reusable scope.

```ruby
Spree::PaymentMethod.joins(:distributors).where(enterprises: {id: Enterprise.all}).count
# => 37
Spree::PaymentMethod.count
# => 25
```

I actually have no idea if a subquery will perform better for this, but I guess that the difference is neglible and find this a lot more maintainable.

Reply:
I was actually talking about the two queries involved here. These are the ones that get generated:

```sql
irb(main):002:0> Spree::PaymentMethod.for_distributors(distributor)
  Spree::PaymentMethod Load (0.8ms)  SELECT "spree_payment_methods".* FROM "spree_payment_methods" INNER JOIN "distributors_payment_methods" ON "distributors_payment_methods"."payment_method_id" = "spree_payment_methods"."id" INNER JOIN "enterprises" ON "enterprises"."id" = "distributors_payment_methods"."distributor_id" WHERE "enterprises"."id" = 2
  Spree::Calculator Load (0.8ms)  SELECT "spree_calculators".* FROM "spree_calculators" WHERE "spree_calculators"."calculable_id" = 13 AND "spree_calculators"."calculable_type" = 'Spree::PaymentMethod' LIMIT 1
  Spree::PaymentMethod Load (0.5ms)  SELECT "spree_payment_methods".* FROM "spree_payment_methods" WHERE "spree_payment_methods"."deleted_at" IS NULL AND "spree_payment_methods"."id" IN (13)
  Spree::Calculator Load (0.3ms)  SELECT "spree_calculators".* FROM "spree_calculators" WHERE "spree_calculators"."calculable_id" = 13 AND "spree_calculators"."calculable_type" = 'Spree::PaymentMethod' LIMIT 1
=> [#<Spree::PaymentMethod::Check id: 13, type: "Spree::PaymentMethod::Check", name: "Cheque (Gippsland Poultry)", description: nil, active: true, environment: "development", created_at: "2018-12-11 16:45:07", updated_at: "2018-12-11 16:45:07", deleted_at: nil, display_on: nil>]
```

We fire 4 of them while the first one already does the job. If you look at it, this one loads the payment methods and then we use the ids of the returned rows to fetch them again in the last one. So, one useless roundtrip to the database.

The calculator is something AR loads because it's an association of that payment method (I didn't expect to see it here)

Reply:
I'm not sure if I'm just missing your point @sauloperez, or possibly also the other way around.

Were you recommending that we should use a join directly, like simply `joins(:distributors).where(enterprises: {id: Enterprise.all})`? My concern is that this could lead to duplicates, like in my example above/below where the result has more than there are total records. It happens when there is a payment method that matches more than one of the enterprises provided.

```ruby
Spree::PaymentMethod.joins(:distributors).where(enterprises: {id: Enterprise.all}).count
# => 37
Spree::PaymentMethod.count
# => 25
```

There is also the option of using `group` or `distinct`, but these would make the named scope not reusable.

So this is one strategy that I think is great, in spite of the extra DB query. Do you disagree?

And yes - it seems even `Spree::PaymentMethod.first` loads the calculator. :sob: I don't want to look into that in this PR.

Reply:
Ok, I get what you did there 👍 We just had a quick chat and agreed on creating a separate issue to fix it differently. We need to start caring for the DB.